### DISCUSSION TITLE
Pushing fvm to the chocolatey

### DISCUSSION THREAD

**Discussion by TarasMazepa** [leoafarias/fvm#777] [General]
How do you prepare and publish fvm for Chocolatey? 

I'm wondering if I can help somehow. Recently, I have automated some of the steps for my own command-line tool. 

One of the things I want to achieve is to have fvm shipped as a binary, so it would not have dart-sdk as a dependency (which forces users to have specific dart-sdk version installed with chocolatey)

**Comment by leoafarias** [comment]
@TarasMazepa thanks for volunteering to help. Right now FVM uses cli-pkg, to automatically deply to chocolatey. You can see it here. https://github.com/leoafarias/fvm/actions/runs/10560273622/job/29254361939 Is there any current issue with Chocolatey, or do you think it can be improved somehow?

**Comment by TarasMazepa** [comment]
Thanks for the response and the links. I have examined the current situation, and here is a run down:

1. In releases, you have already prebuilt packages for different Windows versions https://github.com/leoafarias/fvm/releases/tag/3.2.1

2. Each of them contains a batch file:
```
@echo off
REM This script drives the standalone fvm package, which bundles together a
REM Dart executable and a snapshot of fvm.

set SCRIPTPATH=%~dp0
set arguments=%*
"%SCRIPTPATH%\src\dart.exe" "%SCRIPTPATH%\src\fvm.snapshot" %arguments%
```
3. There is also a portable dart executable and a snapshot of fvm
![image](https://github.com/user-attachments/assets/38d36892-2eb4-4776-ab8e-e006f2ab8b90)
4. If you download `fvm.3.2.1.nupkg` from chocolatey and unpack it, you will see it contains complete source code as well as `fvm.nuspec`, `chocolateyInstall.ps1` and `chocolateyUninstall.ps1`
5. In `fvm.nuspec`, you would notice dart-sdk dependency:
```
    <dependencies>
      <dependency id="dart-sdk" version="[3.4.1]" />
    </dependencies>
```
6. In `chocolateyInstall.ps1` you would see that fvm is being compiled from source code:
```
$ToolsDir = (Split-Path -parent $MyInvocation.MyCommand.Definition)
Write-Host "Fetching Dart dependencies..."
$SourceDir = "$ToolsDir\source"
Push-Location -Path $SourceDir
dart pub get --no-precompile | Out-Null
Pop-Location

New-Item -Path $PackageFolder -Name "bin" -ItemType "directory" | Out-Null
Write-Host "Building executable..."
$ExePath = "$PackageFolder\bin\fvm.exe"
dart compile exe '"-Dversion=3.2.1"' '"-Ddart-version=3.4.1"' "$SourceDir\bin\main.dart" -o $ExePath
Generate-BinFile "fvm" $ExePath
```

Due to that fvm will always decide which dart-sdk user needs to have installed on their machine.

I think this could be improved by removing dependency and shipping prebuild snapshots already generated and attached to the releases. Let me know what do you think.

**Reply by leoafarias** [reply]
We currently use cli_pkg to handle this. Could this be an improvement on that package, or could you suggest a configuration change?

**Reply by TarasMazepa** [reply]
That would be a lot more work. Alternatively I can create separate chocolatey package, for instance `fvmbin`, or `bfvm` which would be deployed with binaries.

**Comment by TarasMazepa** [comment]
@leoafarias, do you have an opinion about what I have described?

**Reply by leoafarias** [reply]
@TarasMazepa, so we would just have a separate .ps1 file? cli_pkg currently handles the submission to Chocolatey. I will also check the changelogs to see if this is something they fixed on their end. Also, it seems there are some Choco issues, which I am not sure if you had experience with, but since you are knowledgeable, I would appreciate your help so we can have a better Chocolatey experience.

**Reply by TarasMazepa** [reply]
@leoafarias here is setup of Chocolatey for commandline tool I'm working on: https://github.com/TarasMazepa/stax/tree/main/chocolatey

Idea is following: I just have these files, and when I have new release I just update url and checksum, I have a GH action which takes case of it: https://github.com/TarasMazepa/stax/blob/main/.github/workflows/release-distribution-trigger.yml#L26

There you can see that it just modifies existing files and creates a PR with changes: https://github.com/TarasMazepa/stax/pull/414

Then I have another GH Action which I can trigger manually to push latest release to the Chocolatey repository: https://github.com/TarasMazepa/stax/blob/main/.github/workflows/chocolatey-release.yml

And one more action to test if Chocolatey is not broken when automatic/manual changes are performed: https://github.com/TarasMazepa/stax/blob/main/.github/workflows/chocolatey-test.yml

I think for fvm we can try doing similar setup:
1. Copy of `chocolatey` folder from my repo: https://github.com/TarasMazepa/stax/tree/main/chocolatey
2. Adding release binary build trigger to modify Url/checksum in nuspec/chocolateyinstall.ps1 files
3. Add GH Action that could be triggered manually to upload latest version to chocolatey (I do this manually and not automatically, as there is manual approval process that takes some time)
